alvr/audio/src/linux.rs

```diff
@@ -0,0 +1,350 @@
+
+    let _receiver = pw_receiver.attach(mainloop.as_ref(), {
+        let mainloop = mainloop.clone();
+        move |_| mainloop.quit()
+    });
+
+    let context = pw::context::Context::new(&mainloop)?;
+    let core = context.connect(None)?;
+
+    let stream = pw::stream::Stream::new(
+        &core,
+        "alvr-audio",
+        pw::properties::properties! {
+            *pw::keys::NODE_NAME => "ALVR Audio",
+            *pw::keys::MEDIA_NAME => "alvr-audio",
+            *pw::keys::MEDIA_TYPE => "Audio",
+            *pw::keys::MEDIA_CATEGORY => "Capture",
+            *pw::keys::MEDIA_CLASS => "Audio/Sink",
+            *pw::keys::MEDIA_ROLE => "Game",
+        },
+    )?;
+
+    let chan_size = std::mem::size_of::<i16>();
+
+    let _listener: StreamListener<i16> = stream
+        .add_local_listener()
+        .process(move |stream, _| match stream.dequeue_buffer() {
+            None => {
+                // Consume value for now because nothing is connected to stream
+                debug!("No pw buffer received")
+            }
+            Some(mut pw_buffer) => {
+                debug!("Starting pw buffer");
+
+                let datas = pw_buffer.datas_mut();
+                if datas.is_empty() {
+                    debug!("Pw buffer empty, skipping process");
+                    return;
+                }
+
+                let pw_data = &mut datas[0];
+                let stride = chan_size * channels_count as usize;
+                let n_frames = (pw_data.chunk().size() / stride as u32) as usize;
+                debug!("n_frames: {}", n_frames);
+                let mut final_buffer: Vec<u8> = Vec::with_capacity(n_frames);
+                if let Some(slice) = pw_data.data() {
+                    debug!("Pw buffer available");
+                    for n_frame in 0..n_frames {
+                        for n_channel in 0..channels_count {
+                            let start = n_frame * stride + (n_channel as usize * chan_size);
+                            let end = start + chan_size;
+                            let channel = &mut slice[start..end];
+                            let slice =
+                                i16::from_ne_bytes(channel.try_into().unwrap()).to_ne_bytes();
+                            final_buffer.extend(slice.iter());
+                        }
+                    }
+                }
+                if !final_buffer.is_empty() && is_running() {
+                    let mut buffer = sender.get_buffer(&()).unwrap();
+                    buffer
+                        .get_range_mut(0, final_buffer.len())
+                        .copy_from_slice(&final_buffer);
+                    debug!("Sending buffer with length of {}", buffer.len());
+                    sender.send(buffer).ok();
+                }
+            }
+        })
+        .register()?;
+
+    let mut audio_info = AudioInfoRaw::new();
+    audio_info.set_format(AudioFormat::S16LE);
+    audio_info.set_rate(44100);
```

Comment:
Rate should not be hardcoded in multiple places

Reply:
will make variable, but tbh for pipeiwre it really doesn't matter

Reply:
i forgot to just get it from that code block with game audio, no need for variable now